File: python/dfml/element.py
from abc import ABC, abstractmethod

class Element(ABC):
	"""
	Abstract base class representing an element in a hierarchical structure.
	
	Attributes:
		NODE (int): Constant representing a node element.
		DATA (int): Constant representing a data element.
		COMMENT (int): Constant representing a comment element.
	"""
	NODE = 0
	DATA = 1
	COMMENT = 2

	@abstractmethod
	def get_element_type(self) -> int:
		"""
		Abstract method to get the type of the element.

		Returns:
			int: An integer representing the type of the element.
		"""
		pass

File: js/main/element.js


/**
 * Base class representing an element in the Dragonfly Markup Language (DFML) structure.
 *
 * @export
 * @class DFMLElement
 */
export class DFMLElement {
  
	/**
	 * Gets the type of the element.
	 * - NODE: 0 - Represents a node.
	 * - DATA: 1 - Represents a data (value only).
	 * - COMMENT: 2 - Represents a comment.
	 * @returns {number} the type of the element.
	 * 
	 * @memberof DFMLElement
	 */
	getElementType() {
		throw new Error('getElementType() must be implemented by subclass');
	}

	/**
	 * Constant representing a DFMLNode element type.
	 *
	 * @readonly
	 * @static
	 * @memberof DFMLElement
	 */
	static get NODE() {
		return 0;
	}
  
	/**
	 * Constant representing a DFMLData.
	 *
	 * @readonly
	 * @static
	 * @memberof DFMLElement
	 */
	static get DATA() {
		return 1;
	}
  
	/**
	 * Constant representing a DFMLComment element type.
	 *
	 * @readonly
	 * @static
	 * @memberof DFMLElement
	 */
	static get COMMENT() {
		return 2;
	}
  }
